### ISSUE TITLE
Questions/Issues using v2.11.0 with Spring Boot and Identity Service

### ISSUE THREAD

**Issue by mr-flannery** [SAP/cloud-security-xsuaa-integration#678]
I am currently testing v2.11.0 of the security libs for multi-tenant validation of Identity Service tokens on CF and K8s.

Good news first, the functionality in general is working as expected! However, I did notice some things.

I was able to use the lib with a `WebFilter` setup like this:

```java
@WebFilter("/api/*")
public class AuthenticationFilter extends GenericFilterBean {

  private final IasTokenAuthenticator tokenAuthenticator;

  public AuthenticationFilter() {
    tokenAuthenticator = new IasTokenAuthenticator();
  }

  @Override
  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
      throws IOException, ServletException {
    try {
       TokenAuthenticationResult authenticationResult = tokenAuthenticator.validateRequest(request, response);
       if (authenticationResult.isAuthenticated()) {
           chain.doFilter(request, response);
       } else {
           // ...
       }
    } finally {
      SecurityContext.clearToken();
    }
  }
}
```

This is all well and good, but I have another application in which I need the `IasTokenAuthenticator` to be autowired by Spring:

```java
  private final IasTokenAuthenticator tokenAuthenticator;

  @Autowired
  public AuthenticationFilter(IasTokenAuthenticator tokenAuthenticator) {
    this.tokenAuthenticator = tokenAuthenticator;
  }
```

This doesn't work out of the box since the `IasTokenAuthenticator` is not a bean and hence cannot be autowired.
This can be worked around like this:

```java
@Component
public class CustomIasTokenAuthenticator extends IasTokenAuthenticator {

  @Override
  protected OAuth2ServiceConfiguration getServiceConfiguration() {
    return super.getServiceConfiguration();
  }

}
```

But it would be nicer if this wouldn't be required (since it e.g. may lead to 

Second, this setup is inconvenient for running the app locally. Locally, the app defaults to using the `CFEnvironment` that tries to access environment properties using `System::getenv`. 
Spring Boot allows maintaining [app configuration](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config) in `application.yml` files. This is easier to maintain in a project than environment variables, since the file can be added to git.
Problem is that `System::getenv` and therefore the `CFEnvironment` used locally cannot read the properties defined in the `application.yml` file. 

Finally, I was looking through the samples in this repository to see whether there's also a way to configure everything declaratively using a `@Configuration` class, instead of wiring it up manually. The only that looked like it is supposed to work with the Identity Service instead of the XSUAA is the [spring-security-hybrid-usage](https://github.com/SAP/cloud-security-xsuaa-integration/blob/main/samples/spring-security-hybrid-usage/src/main/java/sample/spring/security/SecurityConfiguration.java) sample. However, this did not work for me and the sample itself does not compile when I clone the repo and try to build it.

Building it the way it is, it fails with:

```
Description:
Field authConverter in sample.spring.security.SecurityConfiguration required a bean of type 'org.springframework.core.convert.converter.Converter' that could not be found.

The injection point has the following annotations:
        - @org.springframework.beans.factory.annotation.Autowired(required=true)

Action:
Consider defining a bean of type 'org.springframework.core.convert.converter.Converter' in your configuration.
```

I also tried it without the `Converter`, since I'm currently not interested in the token exchange, but that also failed:

```
Description:
Method springSecurityFilterChain in org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration required a bean of type 'org.springframework.security.oauth2.jwt.JwtDecoder' that could not be found.

Action:
Consider defining a bean of type 'org.springframework.security.oauth2.jwt.JwtDecoder' in your configuration.
```

In summary, here are my questions:
1. What is the recommended way to setup the library for usage with the Identity Service in a Spring Boot application?
2. Is there a setup where I can autowire the necessary classes out-of-the-box without adding wrapper classes?
3. Is there a setup where I the lib can read from `application.yml` when running locally?

Thanks!

**Comment by nenaraab** [comment]
Hi @liga-oz 
can you please pick that issue?
Thanks in advance!

**Comment by liga-oz** [comment]
Hi @mr-flannery,

sorry for late reply.

1. and 2. for spring applications I would suggest to use `spring-security` library instead of `java-security`, where you can autowire `IasJwtDecoder` and use `decode()` method that will perform the token validation. `IasJwtDecoder` is also included in `Autoconfiguration` so you can use it out of the box. You can read more about it [here](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/spring-security#auto-configuration)
3. `spring-security` library does support reading application properties files, if you add the `application.yml` file under the resource folder of your app and run it locally it should use credentials from there. 

Regarding the issue with `spring-security-hybrid-sample` I presume you were trying to run it locally? It failed because it didn't find configuration properties, try to add `application.yml` file in the `resources` folder and fill in the the following attributes:

```yaml
sap:
  security:
    services:
      xsuaa:
        xsappname: xsapp!t0815
        uaadomain: localhost
        clientid: sb-clientId!t0815
        clientsecret: pwd
        url: http://localhost

      identity:
        clientid: sb-clientId!t0815
        domains: localhost
```

Hope that clarifies your questions.

Kind Regards,
Liga

**Comment by mr-flannery** [comment]
Hi @liga-oz,

thanks for the reply, I was able to make progress. After adding the configuration to my `application.yml`, my application now starts correctly.

Unfortunately, I did run into more issues:

1. This one is rather minor, but the `sap.security.services.identity.domains` property in the `application.yml` is expected to be a string (as provided in your example). However, in the service bindings to the Identity Service, this property is an array of string. I also took a look at the `JwtIssuerValidator` and domains is also modelled as a `List<String>` there. Consequently, I'd expect the configuration to take a list of strings instead of a single string.
2. More importantly, while the token validation works now, accessing the token does not work for me. I used `@AuthenticationPrincipal Token token` as described in your example, but the token was always `null`. I also applied the config for my Identity Service binding to the example repo, started the app and sent a request to `/sayHello` using a token I fetched via Postman, and I'm running into the same issue here. When I attach a debugger to the controller and manually access `SecurityContextHolder.getContext()`, I do get a result, though.

**Comment by liga-oz** [comment]
Hi @mr-flannery,

1. You can also provide list of strings like so:
      ```yml
            identity:
              clientid: sb-clientId!t0815
              domains: localhost, localhost2
      ```
      both approaches will work and will be converted to String List internally.
2. If the Request was authenticated successfully you should have the token stored in SpringSecurityContext and should be able to access it using the `@AuthenticationPrincipal` Have you changed anything else in the `SecurityConfiguration` of the sample?

**Comment by mr-flannery** [comment]
Hi @liga-oz 

1. Good to know that this works. I do have to ask, though: is this something that you control? The [yaml spec](https://yaml.org/spec/1.2.2/#21-collections) defines the syntax for arrays (collections) differently. I believe that a generic yaml parser will interpret `domains: localhost, localhost2` as string, not as arrays.
2. I have, in fact. I removed the converter stuff since I don't have an XSUAA binding. Here's my config:
```
@Configuration
@EnableWebSecurity(debug = true) // TODO "debug" may include sensitive information. Do not use in a production system!
@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true, jsr250Enabled = true)
@PropertySource(factory = IdentityServicesPropertySourceFactory.class, ignoreResourceNotFound = true, value = { "" })
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

	@Override
	protected void configure(HttpSecurity http) throws Exception {
		// @formatter:off
		http
			.sessionManagement()
				.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
			.and()
				.authorizeRequests()
				// .antMatchers("/sayHello").hasAuthority("Read")
				.antMatchers("/*").authenticated()
				.anyRequest().denyAll()
			.and()
				.oauth2ResourceServer()
				.jwt();
		// @formatter:on
	}
}
````

**Comment by liga-oz** [comment]
Hi @mr-flannery,

1. Yml parsing is done by Spring Framework and it for whatever reason accepts comma separated string list as well as `-` defined array as per the yml spec. So I would say if you want to be on safe side you can define the array as in the yml spec
2. You still need the converter to parse the Jwt into compatible `Principal` type for the `Token` class used in the argument beside the `@AuthenticationPrincipal` and to set the authorities correctly, otherwise groups claim is not parsed as authorities. 
As you're not using Xsuaa you can simplify the converter like so:
```java
class MyCustomTokenAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {

	public AbstractAuthenticationToken convert(Jwt jwt) {
		Collection<GrantedAuthority> groupAuthorities = new ArrayList<>();
		jwt.getClaimAsStringList(TokenClaims.GROUPS).forEach(group -> groupAuthorities.add(new SimpleGrantedAuthority(group)));
		return new AuthenticationToken(jwt, groupAuthorities);
	}
}
```

Kind Regards,
Liga

**Comment by mr-flannery** [comment]
Hi again @liga-oz,

I've made some progress again, here's where I'm at now:

1. regarding the `Converter`, I was able to make it work locally. However, I needed to add the following to my config for it to work:
```
sap:
  security:
    services:
      xsuaa:
        xsappname: DUMMY
        clientid: DUMMY
```
So even though I'm not using an XSUAA instance, the config is still necessary for the application to start in the first place. This shouldn't be necessary.
2. Consequently, I needed to created an XSUAA instance to deploy the app to CF. Again, shouldn't be necessary, ideally.
3. I ran into a strange issue when comparing how my code works locally vs. how it works on CF. Good news first, `@AuthenticationPrincipal Token token` works. However, we have some other places in the code where we want to access the token:
```
final Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
final Object details = SecurityContextHolder.getContext().getAuthentication().getDetails();

System.out.println(String.format("principal is instance of SapIdToken: %b", principal instanceof SapIdToken));
System.out.println(String.format("details is instance of SapIdToken: %b", details instanceof SapIdToken));

final Object thingToCast = principal instanceof SapIdToken ? principal : details instanceof SapIdToken ? details : null;
``` 
When running the app locally, `getPrincipal` returns an instance of `SapIdToken`. On CF, `getDetails` returns an instance of `SapIdToken`.
4. I cannot get it working on K8s. As on CF, I added an XSUAA instance:
```yml
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: mdi-proxy-xsuaa
spec:
  clusterServiceClassExternalName: xsuaa
  clusterServicePlanExternalName: application
  parameters: # I tried it with and without parameters, didn't make a difference
    xsappname: webshop-mdi-proxy
    tenant-mode: shared
    oauth2-configuration:
      credential-types:
        - x509

---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: mdi-proxy-xsuaa-binding
spec:
  instanceRef:
    name: mdi-proxy-xsuaa

---
apiVersion: apps/v1
kind: Deployment
# ...
          volumeMounts:
            - name: mdi-proxy-ias-binding-volume
              mountPath: /etc/secrets/sapbtp/identity/webshop-ias
            - name: mdi-proxy-xsuaa-binding-volume
              mountPath: /etc/secrets/sapbtp/xsuaa/mdi-proxy-xsuaa
      volumes:
        - name: mdi-proxy-ias-binding-volume
          secret:
            secretName: mdi-proxy-ias-binding
        - name: mdi-proxy-xsuaa-binding-volume
          secret:
            secretName: mdi-proxy-xsuaa-binding
```
The problem is again that the `Converter` bean cannot be found, so apparently the XSUAA credentials cannot be read. However, I have no idea what's causing this, since (unless I totally missed something), the instance binding is exactly as described in the sample and as we've done it without previous bindings.

**Comment by mr-flannery** [comment]
Hi @liga-oz ,

thanks again for your help in making it run. I have finished adding this to our Spring Boot project now, i.e. it runs locally, on CF, on K8s, and the tests are working.

Some hints for those that come after me (maybe this can also make its way into your sample, an Identity Service-only sample might be a good idea):

The complete security config should look somewhat like this:

```java
@Configuration
@EnableWebSecurity(debug = true) // TODO "debug" may include sensitive information. Do not use in a production system!
@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true, jsr250Enabled = true)
@PropertySource(factory = IdentityServicesPropertySourceFactory.class, ignoreResourceNotFound = true, value = { "" })
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

	@Override
	protected void configure(HttpSecurity http) throws Exception {
		// @formatter:off
		http
			.sessionManagement()
				.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
			.and()
				.authorizeRequests()
				// .antMatchers("/sayHello").hasAuthority("Read")
				.antMatchers("/*").authenticated()
				.anyRequest().denyAll()
			.and()
				.oauth2ResourceServer()
				.jwt()
				.jwtAuthenticationConverter(new MyCustomTokenAuthenticationConverter());
		// @formatter:on
	}

	class MyCustomTokenAuthenticationConverter implements Converter<Jwt, AbstractAuthenticationToken> {

		public AbstractAuthenticationToken convert(Jwt jwt) {
			final List<String> groups = jwt.getClaimAsStringList(TokenClaims.GROUPS);
			final List<GrantedAuthority> groupAuthorities =
					groups == null
							? Collections.emptyList()
							: groups.stream().map(SimpleGrantedAuthority::new).collect(Collectors.toList());
			return new AuthenticationToken(jwt, groupAuthorities);
		}
	}
}
```

**TL;DR:** Without an XSUAA service, DO NOT use an `@Autowired converter`. This is why I needed an XSUAA binding, even though our app does not use the XSUAA service.

There are two issues leftover for me:
1. I believe there should be a more helpful error message when autowiring dependencies fails due to missing Identity/XSUAA service credentials. Currently, the message only says "could not find a bean XYZ", which usually is an issue of a missing dependency or missing bean-specific configuration. Without knowing what the issue is, it is unlikely to find out that it's caused by missing service credentials.
2. In our security config, we added custom handlers for HTTP 401:
    ```java
    http.exceptionHandling().authenticationEntryPoint(new CustomAuthenticationEntryPoint());
    ```
This handler does still work when I try to access the app without a JWT. However, if the security lib rejects the JWT (invalid issuer, expired, etc.), the error is not correctly propagated to the handler anymore. Do you have an idea what could cause this issue?
